### PR TITLE
Custom Cloudwatch alarms

### PR DESCRIPTION
**Pull request by jamengual**
## what
* Add ability to customize alarms for autoscaling
* Add ability to disable default cloudwatch alarms

## why
* sometimes EC2 instances need to scale base on ECS, ALB, or another type of cloudwatch alarms

### CODE REVIEW

File: autoscaling.tf [hunk 1]
```diff
@@ -24,40 +24,58 @@ resource "aws_autoscaling_policy" "scale_down" {
...
-  comparison_operator = "LessThanOrEqualToThreshold"
-  evaluation_periods  = var.cpu_utilization_low_evaluation_periods
-  metric_name         = "CPUUtilization"
-  namespace           = "AWS/EC2"
-  period              = var.cpu_utilization_low_period_seconds
-  statistic           = var.cpu_utilization_low_statistic
-  threshold           = var.cpu_utilization_low_threshold_percent
-
+resource "aws_cloudwatch_metric_alarm" "all_alarms" {
+  for_each                  = module.this.enabled ? local.all_alarms : null
```
Comment (lorengordon) [inline autoscaling.tf:63]:
```suggestion
  for_each                  = module.this.enabled ? local.all_alarms : {}
```

File: variables.tf [hunk 2]
```diff
@@ -383,3 +383,25 @@ variable "cpu_utilization_low_statistic" {
...
+    treat_missing_data        = string
+    ok_actions                = list(string)
+    insufficient_data_actions = list(string)
+    dimensions_name           = string
+    dimensions_target         = string
+    alarm_description         = string
+    alarm_actions             = list(string)
+  }))
+  default     = {}
+  description = "List of custom CloudWatch alarms configurations"
```
Comment (lorengordon) [inline variables.tf:406]:
```suggestion
  description = "Map of custom CloudWatch alarms configurations"
```

**Review by lorengordon — COMMENTED** [review]
Just a couple quick comments while I'm trying to understand the test failure...

**Review by osterman — APPROVED** [review]
LGTM, @aknysh should probably sign off too

File: variables.tf [hunk 3]
```diff
@@ -383,3 +383,30 @@ variable "cpu_utilization_low_statistic" {
   default     = "Average"
   description = "The statistic to apply to the alarm's associated metric. Either of the following is supported: `SampleCount`, `Average`, `Sum`, `Minimum`, `Maximum`"
 }
+
+variable "default_alarms" {
```
Comment (aknysh) [inline variables.tf:387]:
```suggestion
variable "default_alarms_enabled" {
```

File: variables.tf [hunk 4]
```diff
@@ -383,3 +383,30 @@ variable "cpu_utilization_low_statistic" {
...
+  type        = bool
+  default     = true
+  description = "Enable or disable cpu and memory Cloudwatch alarms"
+}
+variable "custom_alarms" {
```
Comment (aknysh) [inline variables.tf:392]: put a new line above

File: autoscaling.tf [hunk 5]
```diff
@@ -24,40 +24,58 @@ resource "aws_autoscaling_policy" "scale_down" {
...
-  comparison_operator = "LessThanOrEqualToThreshold"
-  evaluation_periods  = var.cpu_utilization_low_evaluation_periods
-  metric_name         = "CPUUtilization"
-  namespace           = "AWS/EC2"
-  period              = var.cpu_utilization_low_period_seconds
-  statistic           = var.cpu_utilization_low_statistic
-  threshold           = var.cpu_utilization_low_threshold_percent
-
+resource "aws_cloudwatch_metric_alarm" "all_alarms" {
+  for_each                  = module.this.enabled && var.default_alarms ? local.all_alarms : null
```
Comment (aknysh) [inline autoscaling.tf:63]: tjis will not enable/disable just the default alarms, but all of them, no?

File: autoscaling.tf [hunk 6]
```diff
@@ -24,40 +24,58 @@ resource "aws_autoscaling_policy" "scale_down" {
...
+      dimensions_name     = "AutoScalingGroupName"
+      dimensions_target   = join("", aws_autoscaling_group.default.*.name)
+      alarm_description   = "Scale down if the CPU utilization is below ${var.cpu_utilization_low_threshold_percent} for ${var.cpu_utilization_low_period_seconds} seconds"
+      alarm_actions       = [join("", aws_autoscaling_policy.scale_down.*.arn)]
+    }
   }
-
-  alarm_description = "Scale up if CPU utilization is above ${var.cpu_utilization_high_threshold_percent} for ${var.cpu_utilization_high_period_seconds} seconds"
-  alarm_actions     = [join("", aws_autoscaling_policy.scale_up.*.arn)]
+  all_alarms = merge(var.custom_alarms, local.default_alarms)
```
Comment (aknysh) [inline autoscaling.tf:59]:
```suggestion
  all_alarms = merge(local.default_alarms, var.custom_alarms)
```

Let's allow the callers to override the default alarms

File: autoscaling.tf (hunk 5 above)
Comment (aknysh) [reply autoscaling.tf:63]: let's do it something like this:

```
locals {
  default_alarms = var.default_alarms_enabled ? {....} : {}
```

then if default alarms are disabled, an empty map will be merged into `local.all_alarms`

**Review by aknysh — DISMISSED** [review]
thanks @jamengual 
LGTM, a few suggestions

File: autoscaling.tf (hunk 1 above)
Comment (lorengordon) [reply autoscaling.tf:64]: @jamengual looks like this false condition is still not compatible with `for_each`. Is there a test case for when `module.this.enabled == false`?

Comment (jamengual) [reply autoscaling.tf:64]: ohhh....let me check

Comment (jamengual) [reply autoscaling.tf:64]: I had to change things around

File: variables.tf (hunk 4 above)
Comment (jamengual) [reply variables.tf:393]: I have no clue why this file is not getting updated

**Review by maximmi — APPROVED** [review]

**Pull request merged**